### PR TITLE
[SPARK-50378][SS] Add custom metric for tracking spent for proc initial state in transformWithState

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add custom metric for tracking spent for proc initial state in transformWithState

### Why are the changes needed?
Adds tracking for time spent in populating initial state

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests

```
[info] Run completed in 2 minutes, 38 seconds.
[info] Total number of tests run: 22
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 22, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @HeartSaVioR - PTAL, thx !

**Review by HeartSaVioR — APPROVED** [review]
+1
Will we have a separate PR to address this in PySpark impl? I guess the execution model may not be exactly the same, but if feasible, better to be consistent.

**Comment by HeartSaVioR** [comment]
I'll leave it to you; I won't block this PR for consistency.

Thanks! Merging to master.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala
Comment (LuciferYang) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:399]: This assertion seems flaky:
- https://github.com/apache/spark/actions/runs/12069501936/job/33656903689
- https://github.com/apache/spark/actions/runs/12060929700/job/33632163805

image

Comment (LuciferYang) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala:399]: The new failed link i n change pipeline:
- https://github.com/apache/spark/actions/runs/12192978771/job/34014487080
- https://github.com/apache/spark/actions/runs/12191406660/job/34010259589
- https://github.com/apache/spark/actions/runs/12185951744/job/33993383845
- https://github.com/apache/spark/actions/runs/12164293819/job/33925652391

image

**Pull request closed without merging**
